### PR TITLE
unify blob client options

### PR DESCRIPTION
**Pull request by Jinming-Hu**

### CODE REVIEW

**Comment by Jinming-Hu** [comment]
@katmsft Please fix build errors in datalake service

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_container_client.hpp [hunk 1]
```diff
@@ -313,8 +313,12 @@ namespace Azure { namespace Storage { namespace Blobs {
   private:
     explicit BlobContainerClient(
         Azure::Core::Http::Url containerUri,
-        std::shared_ptr<Azure::Core::Http::HttpPipeline> pipeline)
-        : m_containerUrl(std::move(containerUri)), m_pipeline(std::move(pipeline))
+        std::shared_ptr<Azure::Core::Http::HttpPipeline> pipeline,
+        Azure::Core::Nullable<EncryptionKey> customerProvidedKey,
```
Comment (katmsft) [inline sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_container_client.hpp:317]: `Nullable` parameters cannot be ignored when being called. Consider `Azure::Core::Nullable<EncryptionKey> customerProvidedKey = Azure::Core::Nullable<EncryptionKey>()`.

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_container_client.hpp [hunk 2]
```diff
@@ -313,8 +313,12 @@ namespace Azure { namespace Storage { namespace Blobs {
...
+        Azure::Core::Nullable<std::string> encryptionScope)
```
Comment (katmsft) [inline sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_container_client.hpp:318]: Ditto.

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_container_client.hpp (hunk 1 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_container_client.hpp:317]: This is a **private** constructor for internal use only.

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_container_client.hpp (hunk 2 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_container_client.hpp:318]: ditto

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_container_client.hpp (hunk 1 above)
Comment (katmsft) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_container_client.hpp:317]: Should every caller that calls this function pass in those values? If not, I'm OK with the change.

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_container_client.hpp:317]: > Should every caller that calls this function pass in those values?

Yes, that's why it is in parameter list without any default value.

Comment (katmsft) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_container_client.hpp:317]: I don't think the statement is valid. DataLake service does not support this feature. Checked .NET/JS. The parameters are meaningless at least for DataLake service.

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_container_client.hpp:317]: > I don't think the statement is valid.

which statement specifically?

> DataLake service does not support this feature

which feature?

Comment (katmsft) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_container_client.hpp:317]: > > Should every caller that calls this function pass in those values?
> 
> Yes, that's why it is in parameter list without any default value.

This statement.

DataLake does not support CPK, so when calling this API, DataLake should not be aware of what it is.

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_container_client.hpp:317]: Fine, but I don't think adding default value for those two parameters actually solves this problem.

I do think the caller should be aware of every parameter of the functions it calls.

Comment (katmsft) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_container_client.hpp:317]: > I do think the caller should be aware of every parameter of the functions it calls.
I think we can agree to disagree. :)

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_container_client.hpp:317]: > > I do think the caller should be aware of every parameter of the functions it calls.
> 
> I think we can agree to disagree. :)

If a blob container has enabled encryption-scope, then it's not accessible anymore to DataLake service. This doesn't make any sense. So it's foreseeable that DataLake service is also going to support encryption scope to improve interoperaibility. Based on this premise, my proposal would be DataLake calls this constructor with `Azure::Core::Nullable<EncryptionKey>()`, which will reduce our workload once these encryption stuff is supported in DataLake.

Comment (katmsft) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_container_client.hpp:317]: > which will reduce our workload once these encryption stuff is supported in DataLake.

If we make them optional with default value at the moment, once datalake support it default value should be removed, then compile error would tell us that DataLake should also be changed to put them in from client options. I think this approach would be better tbh.

Anyway, feel free to leave it as is, not a major concern.

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_container_client.hpp:317]: I don't see why there will be compiler error

**Review by katmsft — APPROVED** [review]

**Pull request merged**